autogpt_platform/backend/backend/server/routers/integrations.py

```diff
@@ -80,20 +81,63 @@ async def callback(
 ) -> CredentialsMetaResponse:
-    handler = _get_provider_oauth_handler(request, provider)
+    logger.debug(f"Received callback for provider: {provider}")
+    try:
+        handler = _get_provider_oauth_handler(request, provider)
+    except Exception as e:
+        logger.error(f"Error getting OAuth handler for provider {provider}: {str(e)}")
+        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")
 
     # Verify the state token
-    if not await store.verify_state_token(user_id, state_token, provider):
-        raise HTTPException(status_code=400, detail="Invalid or expired state token")
+    try:
+        if not await store.verify_state_token(user_id, state_token, provider):
+            logger.warning(f"Invalid or expired state token for user {user_id}")
+            raise HTTPException(
+                status_code=400, detail="Invalid or expired state token"
+            )
+    except Exception as e:
+        logger.error(f"Error verifying state token: {str(e)}")
+        raise HTTPException(
+            status_code=500, detail=f"Error verifying state token: {str(e)}"
+        )
 
     try:
-        credentials = handler.exchange_code_for_tokens(code)
+        scopes = await store.get_scopes_from_state_token(user_id, state_token, provider)
+        logger.debug(f"Retrieved scopes from state token: {scopes}")
+
+        # If scopes are empty, use the default scopes for the provider
+        if not scopes and provider == "google":
+            logger.debug(f"Using default scopes for provider {provider}")
+            scopes = handler.DEFAULT_SCOPES
+
+        credentials = handler.exchange_code_for_tokens(code, scopes)
+        logger.debug(f"Received credentials with final scopes: {credentials.scopes}")
+
+        # Check if the granted scopes are sufficient for the requested scopes
+        if not set(scopes).issubset(set(credentials.scopes)):
+            logger.warning(
+                f"Granted scopes {credentials.scopes} do not include all requested scopes {scopes}"
+            )
+            # You can choose to raise an exception here or handle it in another way
+            # For now, we'll just log the warning and continue
+
     except Exception as e:
-        logger.warning(f"Code->Token exchange failed for provider {provider}: {e}")
-        raise HTTPException(status_code=400, detail=str(e))
+        logger.error(f"Code->Token exchange failed for provider {provider}: {str(e)}")
+        raise HTTPException(
+            status_code=400, detail=f"Failed to exchange code for tokens: {str(e)}"
+        )
 
-    # TODO: Allow specifying `title` to set on `credentials`
-    store.add_creds(user_id, credentials)
+    try:
+        store.add_creds(user_id, credentials)
+    except Exception as e:
+        logger.error(f"Failed to store credentials for user {user_id}: {str(e)}")
+        raise HTTPException(
+            status_code=500, detail=f"Failed to store credentials: {str(e)}"
+        )
```

Comment:
I would prefer to not add individual try/except and log statements for every operation if there isn't a reasonable probability that it will actually break. This will really blow up our codebase for no significant added value. We already have top-level error logging afaik, which is good enough for stuff like this without an expected failure mode.

Reply:
Each of these steps broke so I added checks. This is probably only ever going to trigger for people writing the integrations though

Reply:
> or were suspected

My philosophy: if you don't *know* that it can break in a non-dev scenario and why, don't add a try/except. Otherwise - if we want to be consistent - we'll balloon our codebase by +40% putting try/excepts everywhere.

Reply:
it should Mostly not break if you're not a dev. This just helps push the error to the frontend

if you do break it its because you declined scopes, didn't approve etc which this will handle

Reply:
fixed